### PR TITLE
feat(test): support per test extern files.

### PR DESCRIPTION
**Pull request by rkirov**
The convention is as follows:
test.js will be compiled with test.externs.js (if it exists)
test_with_platform.js will be compiler with platform externs (and
additionally test_with_platform.externs.js if that exists).

### CODE REVIEW

**Comment by rkirov** [comment]
In a follow-up, types_externs_with_platform will be split into smaller tests.

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java [hunk 1]
```diff
@@ -35,6 +34,13 @@ public boolean accept(File dir, String name) {
       return name.endsWith(".js");
     }
   };
+
+  public static final FilenameFilter JS_NO_EXTERNS = new FilenameFilter() {
+    @Override
+    public boolean accept(File dir, String name) {
+      return name.endsWith(".js") && !name.endsWith(".externs.js");
+    }
+  };
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: nit space after?

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java [hunk 2]
```diff
@@ -60,6 +69,15 @@ static File getGoldenFile(final File input) {
     return new File(input.getPath().replaceAll("\\.js$", ".d.ts"));
   }
 
+  static String maybeExternFileName(String testFileName) {
+    String possibleFileName = testFileName.replace(".js",".externs.js");
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: nit: space in `, "`

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java [hunk 3]
```diff
@@ -60,6 +69,15 @@ static File getGoldenFile(final File input) {
     return new File(input.getPath().replaceAll("\\.js$", ".d.ts"));
   }
 
+  static String maybeExternFileName(String testFileName) {
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: I use `maybeDoSomething` for stuff that maybe does some side effect, and `getBlerghOrNull` for stuff that gets something or returns `null` if not found. YMMV, but I've found that quite useful.

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java [hunk 4]
```diff
@@ -60,6 +69,15 @@ static File getGoldenFile(final File input) {
...
+    String possibleFileName = testFileName.replace(".js",".externs.js");
+    Path testDir = FileSystems.getDefault().getPath("src", "test", "java");
+    String packageName = DeclarationGeneratorTests.class.getPackage().getName();
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: I feel like I've seen this package code below, extract a method `getFileInPackage`?

File: src/test/java/com/google/javascript/clutz/ProgramSubject.java [hunk 5]
```diff
@@ -26,21 +26,11 @@
...
-      if (name.contains("_externs")) return PLATFORM;
-      return NONE;
-    }
-  }
-
   /** A stripped down version of Closure's base.js for Clutz tests. */
   private static final SourceFile CLUTZ_GOOG_BASE =
       SourceFile.fromFile("src/test/java/com/google/javascript/clutz/base.js", UTF_8);
+  public boolean withPlatform = false;
+  public String externFile = null;
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: `additionalExternFile`? It doesn't replace the platform externs, does it? Or maybe choose some naming that makes that clear.

File: src/test/java/com/google/javascript/clutz/ProgramSubject.java [hunk 6]
```diff
@@ -96,7 +76,7 @@ StringSubject diagnosticStream() {
   }
 
   private String[] parse() throws AssertionError {
-    Options opts = new Options(externs == Externs.NONE);
+    Options opts = new Options(withPlatform == false && externFile == null);
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: could you add a comment on what the parameter does? `new Options(/* include externs */ ...)?

Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: done.

File: src/test/java/com/google/javascript/clutz/ProgramSubject.java (hunk 5 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: went with `extraExternFile` (indeed it is additive on top of platform externs, if the test needs them).

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java (hunk 4 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: done.

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java (hunk 3 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: done. good suggestion.

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java (hunk 2 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: done. Obviously, I have gotten too used to clang-format, need to integrate it into my workflow.

File: src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java (hunk 1 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/DeclarationGeneratorTests.java]: done

**Comment by mprobst** [comment]
LGTM.

**Pull request merged**